db_stress_tool/expected_state.h

```diff
@@ -19,9 +19,181 @@
+    const uint32_t pending_del = GetValuePart(expected_value, PENDING_DEL_MASK);
+    return pending_del != 0;
+  }
+
+  static void SetPendingDel(uint32_t* expected_value) {
+    SetValuePart(expected_value, PENDING_DEL_MASK, PENDING_DEL_MASK);
+  }
+
+  static void ClearPendingDel(uint32_t* expected_value) {
+    ClearValuePart(expected_value, PENDING_DEL_MASK);
+  }
+
+  static bool IsDeleted(uint32_t expected_value) {
+    const uint32_t deleted = GetValuePart(expected_value, DEL_MASK);
+    return deleted != 0;
+  }
+
+  static void SetDeleted(uint32_t* expected_value) {
+    SetValuePart(expected_value, DEL_MASK, DEL_MASK);
+  }
+
+  static void ClearDeleted(uint32_t* expected_value) {
+    ClearValuePart(expected_value, DEL_MASK);
+  }
+
+  static uint32_t GetDelMask() { return DEL_MASK; }
+
+  static bool IsValueBaseValid(uint32_t value_base) {
+    return IsValuePartValid(value_base, VALUE_BASE_MASK);
+  }
+
+ private:
+  // The 32-bit ExpectedValue is divided into following parts:
+  // Bit 0 - 14: value base
+  static constexpr uint32_t VALUE_BASE_MASK = 0x7fff;
+  static constexpr uint32_t VALUE_BASE_DELTA = 1;
+  // Bit 15: whether write to this value base is pending (0 equals `false`)
+  static constexpr uint32_t PENDING_WRITE_MASK = (uint32_t)1 << 15;
+  // Bit 16 - 29: deletion counter (i.e, number of times this value base has
+  // been deleted)
+  static constexpr uint32_t DEL_COUNTER_MASK = 0x3fff0000;
+  static constexpr uint32_t DEL_COUNTER_DELTA = (uint32_t)1 << 16;
+  // Bit 30: whether deletion of this value base is pending (0 equals `false`)
+  static constexpr uint32_t PENDING_DEL_MASK = (uint32_t)1 << 30;
+  // Bit 31: whether this value base is deleted (0 equals `false`)
+  static constexpr uint32_t DEL_MASK = (uint32_t)1 << 31;
+
+  static uint32_t GetValuePart(uint32_t expected_value,
+                               uint32_t value_part_mask) {
+    return expected_value & value_part_mask;
+  }
+
+  static uint32_t GetIncrementedValuePart(uint32_t current_value_part,
+                                          uint32_t value_part_mask,
+                                          uint32_t value_part_delta) {
+    assert(IsValuePartValid(current_value_part, value_part_mask));
+    return GetValuePart(current_value_part + value_part_delta, value_part_mask);
+  }
+
+  static bool IsValuePartValid(uint32_t value_part, uint32_t value_part_mask) {
+    return (value_part & (~value_part_mask)) == 0;
+  }
+
+  static void SetValuePart(uint32_t* expected_value, uint32_t value_part_mask,
+                           uint32_t new_value_part) {
+    assert(expected_value);
+    assert(IsValuePartValid(new_value_part, value_part_mask));
+    ClearValuePart(expected_value, value_part_mask);
+    *expected_value |= new_value_part;
+  }
+
+  static void ClearValuePart(uint32_t* expected_value,
+                             uint32_t value_part_mask) {
+    assert(expected_value);
+    *expected_value &= (~value_part_mask);
+  }
```

Comment:
I would move at least all of these to `ExpectedValue`. Only a couple (`IsValueBaseValid()` and `IsValuePartValid()`) need to remain static; the rest can become member functions.

Reply:
Fixed